### PR TITLE
bpo-45260: Add superinstruction, UNPACK_SEQUENCE__STORE_FAST

### PR DESCRIPTION
**Pull request by zcpara**
https://bugs.python.org/issue45260

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@zcpara

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by gvanrossum** [comment]
You don't need to do this in specialize.c, right? You can do it in the compiler. That's faster.

**Comment by zcpara** [comment]
> You don't need to do this in specialize.c, right? You can do it in the compiler. That's faster.

I implement it in the compiler. The performance of both implementations are almost the same.

File: Lib/opcode.py [hunk 1]
```diff
@@ -247,6 +247,7 @@ def jabs_op(name, op):
     "STORE_ATTR_SLOT",
     "STORE_ATTR_WITH_HINT",
     # Super instructions
+    "UNPACK_SEQUENCE_ST",
```
Comment (Fidget-Spinner) [inline Lib/opcode.py]: Please keep with the naming convention and rename this to ``UNPACK_SEQUENCE__STORE_FAST``. Thank you.

**Review by Fidget-Spinner — COMMENTED** [review]
From [your benchmarks](https://github.com/faster-cpython/ideas/issues/16#issuecomment-925020450) 1.008x faster (0.8% faster) is almost in the realm of noise. It seems that only ``unpack_sequence`` (which is a microbenchmark) benefited. I don't know if we'll ever run out of instructions, but this doesn't seem to worth it IMO.

FWIW, thus far the superinstructions implemented by Mark have been via the specializing interpreter. From what I understand, the main benefit over a compiler implementation is that when users dissassemble/inspect/trace bytecode, they won't encounter some strange-looking instruction and it will instead fall back to the old instruction. So we don't need to document or support third party tools.

**Comment by Fidget-Spinner** [comment]
Nonetheless, I wanted to say that this is still very impressive work! Thank you for implementing the instruction and benchmarking this.

File: Lib/opcode.py (hunk 1 above)
Comment (zcpara) [reply Lib/opcode.py]: Done.

**Comment by gvanrossum** [comment]
> FWIW, thus far the superinstructions implemented by Mark have been via the specializing interpreter. From what I understand, the main benefit over a compiler implementation is that when users dissassemble/inspect/trace bytecode, they won't encounter some strange-looking instruction and it will instead fall back to the old instruction. So we don't need to document or support third party tools.

I don’t want to derail this PR, but I feel in general the compiler should do it if it can, like in this case. “Specialization” would be for cases where runtime data needs to be taken into account (e.g. BINARY_SUBSCR vs. BINARY_SUBSCR_LIST etc.). But we should really see which has the shorter code.

**Comment by gvanrossum** [comment]
> I don’t want to derail this PR, but I feel in general the compiler should do it if it can, like in this case.

Sorry to seem arguing with myself, but actually I stand corrected. The specializer produces a bunch of super-instructions, like LOAD_FAST__LOAD_FAST. See the switch in optimize() in specialize.c.

**Comment by zcpara** [comment]
Update the implementation in the compiler.

**Comment by markshannon** [comment]
We now specialize `UNPACK_SEQUENCE`, which means that adding this superinstruction would interfere with that specialization.
We have experimented with specialized superinstructions and, with the possibe exception of `LOAD_FAST__LOAD_ATTR_INSTANCE_VALUE`, they don't seem to be worthwhile.

**Pull request closed without merging**